lms/djangoapps/completion/models.py

```diff
@@ -73,37 +125,100 @@ def submit_completion(self, user, course_key, block_key, completion):
                 IntegrityError and OperationalError are relatively common
                 subclasses.
         """
+        self.validate(user, course_key, block_key)
+        block_type = block_key.block_type
 
-        # Raise ValueError to match normal django semantics for wrong type of field.
-        if not isinstance(course_key, CourseKey):
-            raise ValueError(
-                "course_key must be an instance of `opaque_keys.edx.keys.CourseKey`.  Got {}".format(type(course_key))
-            )
-        try:
-            block_type = block_key.block_type
-        except AttributeError:
-            raise ValueError(
-                "block_key must be an instance of `opaque_keys.edx.keys.UsageKey`.  Got {}".format(type(block_key))
-            )
+        obj, is_new = self.get_or_create(
+            user=user,
+            course_key=course_key,
+            block_type=block_type,
+            block_key=block_key,
+            defaults={'completion': completion},
+        )
+        if not is_new and obj.completion != completion:
+            obj.completion = completion
+            obj.full_clean()
+            obj.save()
```

Comment:
I tried using `update_or_create` plus the `pre_save` hook here like in b31f337 but got some test errors.

Feel free to nut them out and do the same here if you can.

Reply:
Test errors? Works just fine over here.

Reply:
There is an issue with this fix. A few of the unit tests are failing because it adds extra queries. It added 2 additional queries per call. @jcdyer and I were talking the other day about queries and keeping them as limited as possible on the completion API.

Reply:
I am not sure where you are trying to take the code. This was kind of a large change that I am not sure was necessary. I am thinking it make sense to revert rather than change working code further.

Reply:
I am so sorry!  I didn't intend to commit those changes.  I found what you noted, that it made those tests fail, though I didn't know why.

Reverted with 1dea1fe